src/agentscope/agents/web_voyager_agent.py

```diff
@@ -0,0 +1,428 @@
+        self.memory.add(
+            message_from_dict(
+                {
+                    "role": "system",
+                    "content": self.sys_prompt,
+                    "name": "system",
+                },
+            ),
+        )
+
+        it = 0
+
+        while it < self.max_iter:
+            self.speak(f"Iter: {it}")
+            it += 1
+            if not fail_obs:
+                try:
+                    (
+                        _web_eles,
+                        web_eles_text,
+                        screenshot_bytes,
+                        web_ele_fields,
+                    ) = self.browser.crawl_page(with_meta=True)
+                    web_eles_text = "\n".join(web_eles_text)
+                except Exception as e:
+                    logger.error("Driver error when adding set-of-mark.")
+                    logger.error(e)
+                    break
+
+                # get screnn shot
+                img_path = os.path.join(
+                    self.task_dir,
+                    f"screenshot_{it}.png",
+                )
+
+                file_manager.save_image(screenshot_bytes, img_path)
+
+                curr_msg = self._get_step_msg(
+                    it,
+                    warn_obs,
+                    img_path,
+                    web_eles_text,
+                )
+
+                self.memory.add(curr_msg)
+            else:
+                # TODO change fail obs
+                curr_msg = message_from_dict(
+                    {
+                        "role": "user",
+                        "name": "user",
+                        "content": fail_obs,
+                    },
+                )
+                self.memory.add(curr_msg)
+
+            messages = self._clip_message_and_obs(self.memory.get_memory())
+
+            formated_messages = self.model.format(messages)
+            gpt_4v_res = self.model(formated_messages).text
+            self.speak(gpt_4v_res)
+
+            self.memory.add(
+                message_from_dict(
+                    {
+                        "role": "assistant",
+                        "content": gpt_4v_res,
+                        "name": "assistant",
+                    },
+                ),
+            )
+
+            # extract action info
+            try:
+                assert "Thought:" in gpt_4v_res and "Action:" in gpt_4v_res
+            except AssertionError as e:
+                logger.error(e)
+                fail_obs = "Format ERROR: Both 'Thought' and 'Action' should be included in your reply."  # noqa
+                continue
+
+            # bot_thought = re.split(pattern, gpt_4v_res)[1].strip()
+            bot_action = re.split(pattern, gpt_4v_res)[2].strip()
+            action_key, info = self._extract_action(bot_action)
+
+            fail_obs = ""
+            warn_obs = ""
+            try:
+                # TODO how to prevent bad auto page navigation?
+
+                if action_key == "click":
```

Comment:
- We should carefully consider which module is responsible for providing web browsing functionality, the agent or the WebBrowser class? 
- In the current PR, the functionality is spread across both the agent and WebBrowser class. 
For example, the following function calling process should be contained in the WebBrowser class rather than the agent.

Reply:
Moved to browser class.